--- newlib-2010q1/newlib/libc/include/sys/config.h.orig
+++ newlib-2010q1/newlib/libc/include/sys/config.h
@@ -125,6 +125,11 @@
 #define _REENT_SMALL
 #endif /* __m32c__ */

+/* For github.com/hugovincent/mbed-freertos */
+#if defined(__thumb__) || defined(__thumb2__) || defined(__arm__)
+#define __DYNAMIC_REENT__
+#endif
+
 #ifdef __thumb2__
 /* Thumb-2 based ARMv7M devices are really small.  */
 #define _REENT_SMALL
--- newlib-2010q1/newlib/libc/include/sys/features.h.orig
+++ newlib-2010q1/newlib/libc/include/sys/features.h
@@ -175,6 +175,12 @@ extern "C" {
 #define _POSIX_TIMERS				     1
 #endif
 
+/* For github.com/hugovincent/mbed-freertos */
+#if defined(__thumb__) || defined(__thumb2__) || defined(__arm__)
+#define _POSIX_TIMERS					1
+#define _POSIX_MONOTONIC_CLOCK			200112L
+#endif
+
 #ifdef __cplusplus
 }
 #endif
